transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c

```diff
@@ -0,0 +1,347 @@
+
+
+    submit(env, theControl->ioContext, iocb);
+}
+
+JNIEXPORT jint JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_poll
+  (JNIEnv * env, jobject obj, jobject contextPointer, jobjectArray callbacks, jint min, jint max) {
+
+    int i = 0;
+    struct io_control * theControl = getIOControl(env, contextPointer);
+
+    int result = io_getevents(theControl->ioContext, min, max, theControl->events, 0);
+
+    for (i = 0; i < result; i++) {
+        struct iocb * iocbp = (struct iocb *)theControl->events[i].obj;
+
+        int eventResult = theControl->events[i].res;
+
+        #ifdef DEBUG
+            fprintf (stderr, "Poll res: %d totalRes=%d\n", eventResult, result);
+        #endif
+
+        if (eventResult < 0) {
+            #ifdef DEBUG
+                fprintf (stderr, "Error: %s\n", strerror(-eventResult));
+            #endif
+
+            jstring jstrError = (*env)->NewStringUTF(env, strerror(-eventResult));
+
+            jobject errorObject = (*env)->NewObject(env,
+                                                   theControl->errorInfoClass,
+                                                   theControl->errorInfoConstr,
+                                                   (jobject)(iocbp->data),
+                                                   (jint)(-eventResult),
+                                                   jstrError);
+
+            if (errorObject == NULL) {
+                return -1;
+            }
+
+            (*env)->SetObjectArrayElement(env, callbacks, i, errorObject);
+        }
+        else {
+            (*env)->SetObjectArrayElement(env, callbacks, i, (jobject)iocbp->data);
+        }
+
+        if (iocbp->data != NULL) {
+            (*env)->DeleteGlobalRef(env, (jobject)iocbp->data);
+        }
+
+        putIOCB(theControl, iocbp);
+    }
+}
+
+JNIEXPORT jobject JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_newAlignedBuffer
+(JNIEnv * env, jclass clazz, jint size, jint alignment) {
+    if (size % alignment) {
+        throwRuntimeException(env, "Buffer size needs to be aligned to passed argument");
+        return 0;
+    }
+
+    // This will allocate a buffer, aligned by alignment.
+    // Buffers created here need to be manually destroyed by destroyBuffer, or this would leak on the process heap away of Java's GC managed memory
+    // NOTE: this buffer will contain non initialized data, you must fill it up properly
+    void * buffer = 0;
+    int result = posix_memalign(&buffer, alignment, size);
+
+    if (result) {
+        throwRuntimeException(env, exceptionMessage("Can't allocate posix buffer:", result));
+        return 0;
+    }
+
+    jobject jbuffer = (*env)->NewDirectByteBuffer(env, buffer, size);
+
+    return jbuffer;
+}
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_freeBuffer
+  (JNIEnv * env, jclass clazz, jobject jbuffer) {
+
+  	void *  buffer = (*env)->GetDirectBufferAddress(env, jbuffer);
```

Comment:
Do you need another temporary variable?  Unless you need to check for `NULL` (which isn't being done now) just inline:

```c
free((*env)->GetDirectBufferAddress(env, jbuffer));
```